qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,849 @@
+                constraints.append(_lambdify_ite(expr, params))
+            else:
+                constraints.append(sym.lambdify(params, expr))
+        cls.global_constraints[pulse_name] = constraints
+
+
+class SymbolicPulse(Pulse):
+    """The abstract superclass for parametric pulses with symbolic expression.
+
+    A symbolic pulse subclass can be defined with an envelope and parameter constraints.
+    The parameters constitute the waveform should be defined in the class attribute
+    :attr:`SymbolicPulse.PARAM_DEF`. This must include ``duration`` as a first element
+    to define the length of pulse instruction for scheduling.
+    For now, envelope and constraints must be defined with SymPy expressions.
+
+    .. rubric:: Envelope function
+
+    This is defined with a class method :meth:`_define_envelope` that returns
+    a single symbolic expression which is a function of ``t`` and ``duration``.
+    The time ``t`` and ``duration`` are in units of dt, i.e. system cycle time,
+    and this function is sampled with a discrete time vector in [0, ``duration``]
+    sampling the pulse envelope at every 0.5 dt (middle sampling strategy) when
+    :meth:`get_waveform` method is called.
+    The function can generate complex-valued samples representing a phase in Gaussian plane,
+    and real and imaginary part may be supplied to two quadratures of
+    the IQ mixer in the control electronics.
+    The sample data is not generated until the get method is called thus
+    a symbolic pulse instance only stores parameter values and waveform shape,
+    which greatly reduces memory footprint during the program generation.
+
+
+    .. rubric:: Constraint functions
+
+    This is defined with a class method :meth:`_define_constraints` that returns
+    a list of symbolic expressions. These constraint functions are called
+    when a symbolic pulse instance is created with assigned parameter values with ``limit``
+    representing the policy of allowing the maximum amplitude exceeding 1.0
+    (defaults to ``True``, i.e. it doesn't allow amplitude > 1.0).
+    The symbolic pulse subclass will be instantiated
+    only when all constraint functions return ``True``.
+    This means the return value of these symbolic expressions must be boolean.
+    Any number of expression can be defined in the class method.
+    When branching logic (if clause) is necessary, one can use `SymPy ITE`_ class
+    or a list of expressions consisting of expression for "IF", "THEN", and "ELSE".
+    Typically, it evaluates the maximum pulse amplitude so that it doesn't exceed the
+    signal dynamic range [-1.0, 1.0]. When validation is not necessary,
+    this method can return an empty list.
+
+    .. _SymPy ITE: https://docs.sympy.org/latest/modules/logic.html#sympy.logic.boolalg.ITE
+
+
+    .. rubric:: Examples
+
+    This is an example of how a pulse author can define new pulse subclass.
+
+    .. code-block:: python
+
+        import sympy as sym
```

Comment:
Bake to Jake's comment, is this the right place for the import?

Reply:
This should be okey because this is outside of Qiskit, e.g. jupyter notebook that a user is running. Jake's comment is about dynamics of package import within the Qiskit.